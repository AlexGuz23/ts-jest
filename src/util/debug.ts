import { format } from 'util'
import { resolve } from 'path'
import { appendFileSync } from 'fs'

export let DEBUG_MODE!: boolean

export let debug!: typeof console.log
export let warn!: typeof console.warn

export let wrapWithDebug!: <T extends (...args: any[]) => any>(
  msg: string | string[] | ((...args: any[]) => string | string[]),
  func: T,
) => T

type LogKind = 'log' | 'warn'
type Logger = (kind: LogKind, ...args: any[]) => void

export let LOG_PREFIX = 'ts-jest:'

export const defaultLogger: Logger = (
  kind: LogKind,
  msg: string = '',
  ...args: any[]
) => {
  // we use stderr/stdout dirrectly so that the log won't be swallowed by jest
  // the dummy calback is to ensure output on CI with node 6
  if (kind === 'warn') {
    process.stderr.write(format(msg, ...args) + '\n', () => undefined)
  } else if (kind) {
    process.stdout.write(format(msg, ...args) + '\n', () => undefined)
  }
}

let writeLogToFile: string | undefined
const markedLogFiles: { [path: string]: 0 } = Object.create(null)

const time = () => new Date().toISOString()
const arrify = (
  input: string | string[] | ((...a: any[]) => string | string[]),
  args: any[],
) => {
  const strOrArray = typeof input === 'function' ? input(...args) : input
  return Array.isArray(strOrArray) ? strOrArray : [strOrArray]
}

interface SetupOptions {
  enabled?: boolean
  logger?: Logger
  prefix?: string
  logFile?: string | false
}
export function __setup({
  logger = defaultLogger,
  logFile = process.env.TS_JEST_DEBUG || process.env.TS_JEST_DEBUG_FILE
    ? resolve(
        process.cwd(),
        process.env.TS_JEST_DEBUG_FILE || 'ts-jest-debug.log',
      )
    : undefined,
  enabled = !!process.env.TS_JEST_DEBUG || logger !== defaultLogger,
  prefix = 'ts-jest:',
}: SetupOptions = {}) {
  writeLogToFile = logFile || undefined
  DEBUG_MODE = enabled
  LOG_PREFIX = prefix

  // wrap logger so that we write to file as well
  if (writeLogToFile) {
    logger = (old => (kind: LogKind, ...args: [any, ...any[]]) => {
      const prefix = `[${time()}][${kind}]`
      const message = format(...args)
      appendFileSync(
        writeLogToFile as string,
        `${prefix} ${message.replace(/\n/g, '\\n')}\n`,
        'utf8',
      )
      // call the original logger only if debug enabled
      if (enabled) old(kind, ...args)
    })(logger)
  }

  // log debug
  debug =
    DEBUG_MODE || writeLogToFile
      ? (...args: any[]) => logger('log', LOG_PREFIX, ...args)
      : () => undefined

  // log warning
  warn = (...args: any[]) => logger('warn', LOG_PREFIX, ...args)

  // function wrapper
  wrapWithDebug =
    DEBUG_MODE || writeLogToFile
      ? (msg, func) =>
          function wrapper(this: any) {
            logger('log', LOG_PREFIX, ...arrify(msg, [...arguments]))
            return func.apply(this, arguments)
          } as any
      : (_, func) => func

  // mark log file
  if (writeLogToFile && !(writeLogToFile in markedLogFiles)) {
    markedLogFiles[writeLogToFile] = 0
    debug('initiailized')
  }
}

__setup()
