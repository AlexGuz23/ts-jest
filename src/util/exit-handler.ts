export interface OnExitHandlerOptions {
  cleanExit: boolean
  exitCode?: number
  signal?: string
  error?: Error
}

interface InternalExitHandlerOptions {
  exitCode?: number
  cleanExit?: boolean
  signal?: string
  error?: Error
  exitWithCode?: number
}

let installed: boolean = false
/**
 * @see https://stackoverflow.com/a/14032965/1603618
 * @param callback Called when the program exists
 */
export function install(callback: (opt: OnExitHandlerOptions) => void) {
  if (installed) throw new Error(`On exit already installed.`)
  installed = true

  function exitHandler({
    exitCode,
    error,
    signal,
    cleanExit = false,
    exitWithCode,
  }: InternalExitHandlerOptions = {}) {
    try {
      callback({ error, exitCode, cleanExit, signal })
    } catch (err) {}
    if (exitWithCode) {
      process.exit(exitCode)
    }
  }

  // do something when app is closing
  process.on('exit', exitCode => exitHandler({ exitCode, cleanExit: true }))
  // catches ctrl+c event
  process.on('SIGINT', () => exitHandler({ signal: 'SIGINT', exitWithCode: 1 }))
  //  catches "kill pid" (for example: nodemon restart)
  process.on('SIGUSR1', () =>
    exitHandler({ signal: 'SIGUSR1', exitWithCode: 1 }),
  )
  process.on('SIGUSR2', () =>
    exitHandler({ signal: 'SIGUSR2', exitWithCode: 1 }),
  )
  // catches uncaught exceptions
  process.on('uncaughtException', error => exitHandler({ error }))
}
