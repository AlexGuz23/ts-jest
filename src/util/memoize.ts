const cacheProp = Symbol.for('[memoize]')

export function Memoize(keyBuilder?: (...args: any[]) => any) {
  return (_: object, propertyKey: string, descriptor: TypedPropertyDescriptor<any>) => {
    if (descriptor.value != null) {
      descriptor.value = memoize(propertyKey, descriptor.value, keyBuilder || ((v: any) => v))
    } else if (descriptor.get != null) {
      descriptor.get = memoize(propertyKey, descriptor.get, keyBuilder || (() => propertyKey))
    }
  }
}

function ensureCache<T extends any>(target: T, reset = false): { [key in keyof T]?: Map<any, any> } {
  if (reset || !target[cacheProp]) {
    Object.defineProperty(target, cacheProp, {
      value: Object.create(null),
      configurable: true,
    })
  }
  return target[cacheProp]
}

function ensureChildCache<T extends any>(target: T, key: keyof T, reset = false): Map<any, any> {
  const dict = ensureCache(target)
  if (reset || !dict[key]) {
    dict[key] = new Map<any, any>()
  }
  return dict[key] as Map<any, any>
}

function memoize(
  namespace: string,
  func: (...args: any[]) => any,
  keyBuilder: (...args: any[]) => any,
): (...args: any[]) => any {
  return function(this: any, ...args: any[]): any {
    const cache = ensureChildCache(this, namespace)
    const key = keyBuilder.apply(this, args)
    if (cache.has(key)) return cache.get(key) as any // tslint:disable-line
    const res: any = func.apply(this, args)
    cache.set(key, res)
    return res
  }
}
/* tslint:disable:unified-signatures */
export function clearCache<T extends { [cacheProp]: any }>(target: T): void
export function clearCache<T extends { [cacheProp]: any }>(target: T, key: keyof T): void
export function clearCache<T extends { [cacheProp]: any }>(target: T, key: keyof T, hash: string): void
export function clearCache<T extends { [cacheProp]: any }>(target: T, key?: keyof T, hash?: string): void {
  if (!target[cacheProp]) return
  if (hash) {
    ensureChildCache(target, key as keyof T).delete(hash)
  } else if (key) {
    ensureChildCache(target, key, true)
  } else {
    ensureCache(target, true)
  }
}
