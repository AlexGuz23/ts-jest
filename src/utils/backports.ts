function warn(...data: any[]) {
  console.warn(...data);
}

function warnConfig(fromPath: string, toPath: string, note?: string) {
  warn(
    `"{jest-config}.${fromPath}" is deprecated, use "{jest-config}.${toPath}" instead.\n` +
      `    â†³ ${note}`,
  );
}

export function backportJestConfig<
  T extends jest.InitialOptions | jest.ProjectConfig
>(config: T = {} as any): T {
  const { globals = {} } = config as any;
  const { 'ts-jest': tsJest = {} } = globals as any;
  const mergeTsJest: any = {};

  if ('__TS_CONFIG__' in globals) {
    warnConfig('globals.__TS_CONFIG__', 'globals.ts-jest.tsConfig');
    if (typeof globals.__TS_CONFIG__ === 'object') {
      mergeTsJest.tsConfig = globals.__TS_CONFIG__;
    }
    delete globals.__TS_CONFIG__;
  }

  if ('__TRANSFORM_HTML__' in globals) {
    warnConfig(
      'globals.__TRANSFORM_HTML__',
      'globals.ts-jest.stringifyContentRegex',
    );
    if (globals.__TRANSFORM_HTML__) {
      mergeTsJest.stringifyContentRegex = /\.html?$/;
    }
    delete globals.__TRANSFORM_HTML__;
  }

  if ('tsConfigFile' in tsJest) {
    warnConfig('globals.ts-jest.tsConfigFile', 'globals.ts-jest.tsConfig');
    if (tsJest.tsConfigFile) {
      mergeTsJest.tsConfig = tsJest.tsConfigFile;
    }
    delete tsJest.tsConfigFile;
  }

  if ('useBabelrc' in tsJest) {
    warnConfig(
      'globals.ts-jest.useBabelrc',
      'globals.ts-jest.babelJest',
      'See `babel-jest` related issue: https://github.com/facebook/jest/issues/3845',
    );
    if (tsJest.useBabelrc != null) {
      mergeTsJest.babelJest = tsJest.useBabelrc ? true : {};
    }
    delete tsJest.useBabelrc;
  }

  if ('babelConfig' in tsJest) {
    warnConfig('globals.ts-jest.babelConfig', 'globals.ts-jest.babelJest');
    if (tsJest.babelConfig != null) {
      mergeTsJest.babelJest = tsJest.babelConfig;
    }
    delete tsJest.useBabelrc;
  }

  if ('skipBabel' in tsJest) {
    warnConfig('globals.ts-jest.skipBabel', 'globals.ts-jest.babelJest');
    if (tsJest.skipBabel === false && !mergeTsJest.babelJest) {
      mergeTsJest.babelJest = true;
    }
    delete tsJest.skipBabel;
  }

  return {
    ...(config as any),
    globals: {
      ...globals,
      'ts-jest': {
        ...mergeTsJest,
        ...tsJest,
      },
    },
  };
}
