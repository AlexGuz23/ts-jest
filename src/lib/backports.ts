import { interpolate, Deprecateds } from './messages'
import { warn } from './debug'

export function backportJestConfig<
  T extends jest.InitialOptions | jest.ProjectConfig
>(config: T = {} as any, silent = false): T {
  const { globals = {} } = config as any
  const { 'ts-jest': tsJest = {} } = globals as any
  const mergeTsJest: any = {}
  const warnConfig = silent
    ? () => undefined
    : (oldPath: string, newPath: string, note?: string) => {
        warn(
          interpolate(
            note ? Deprecateds.ConfigOptionWithNote : Deprecateds.ConfigOption,
            {
              oldPath,
              newPath,
              note
            }
          )
        )
      }

  if ('__TS_CONFIG__' in globals) {
    warnConfig('globals.__TS_CONFIG__', 'globals.ts-jest.tsConfig')
    if (typeof globals.__TS_CONFIG__ === 'object') {
      mergeTsJest.tsConfig = globals.__TS_CONFIG__
    }
    delete globals.__TS_CONFIG__
  }

  if ('__TRANSFORM_HTML__' in globals) {
    warnConfig(
      'globals.__TRANSFORM_HTML__',
      'globals.ts-jest.stringifyContentPathRegex'
    )
    if (globals.__TRANSFORM_HTML__) {
      mergeTsJest.stringifyContentPathRegex = '\\.html?$'
    }
    delete globals.__TRANSFORM_HTML__
  }

  if ('tsConfigFile' in tsJest) {
    warnConfig('globals.ts-jest.tsConfigFile', 'globals.ts-jest.tsConfig')
    if (tsJest.tsConfigFile) {
      mergeTsJest.tsConfig = tsJest.tsConfigFile
    }
    delete tsJest.tsConfigFile
  }

  if ('enableTsDiagnostics' in tsJest) {
    warnConfig(
      'globals.ts-jest.enableTsDiagnostics',
      'globals.ts-jest.diagnostics'
    )
    if (tsJest.enableTsDiagnostics) {
      mergeTsJest.diagnostics =
        typeof tsJest.enableTsDiagnostics === 'string'
          ? { pathRegex: tsJest.enableTsDiagnostics }
          : true
    } else {
      mergeTsJest.diagnostics = false
    }
    delete tsJest.enableTsDiagnostics
  }

  if ('useBabelrc' in tsJest) {
    warnConfig(
      'globals.ts-jest.useBabelrc',
      'globals.ts-jest.babelConfig',
      Deprecateds.ConfigOptionUseBabelRcNote
    )
    if (tsJest.useBabelrc != null) {
      mergeTsJest.babelConfig = tsJest.useBabelrc ? true : {}
    }
    delete tsJest.useBabelrc
  }

  // if ('babelConfig' in tsJest) {
  //   warnConfig('globals.ts-jest.babelConfig', 'globals.ts-jest.babelConfig')
  //   if (tsJest.babelConfig != null) {
  //     mergeTsJest.babelConfig = tsJest.babelConfig
  //   }
  //   delete tsJest.babelConfig
  // }

  if ('skipBabel' in tsJest) {
    warnConfig('globals.ts-jest.skipBabel', 'globals.ts-jest.babelConfig')
    if (tsJest.skipBabel === false && !mergeTsJest.babelConfig) {
      mergeTsJest.babelConfig = true
    }
    delete tsJest.skipBabel
  }

  return {
    ...(config as any),
    globals: {
      ...globals,
      'ts-jest': {
        ...mergeTsJest,
        ...tsJest
      }
    }
  }
}
