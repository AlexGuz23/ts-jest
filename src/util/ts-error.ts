import { inspect } from 'util'
import { BaseError } from 'make-error'
import { Helps, Errors, interpolate } from './messages'
import { debug } from './debug'

/**
 * @internal
 */
export const INSPECT_CUSTOM = inspect.custom || 'inspect'

/**
 * TypeScript diagnostics error.
 */
export class TSError extends BaseError {
  name = 'TSError'

  constructor(public diagnosticText: string, public diagnosticCodes: number[]) {
    super(
      interpolate(Errors.UnableToCompileTypeScript, {
        diagnostics: diagnosticText.trim(),
        help: Helps.IgnoreDiagnosticCode,
      }),
    )
    debug('TSError#constructor', diagnosticCodes, diagnosticText)
    // ensure we blacklist any of our code
    Object.defineProperty(this, 'stack', { value: '' })
  }

  /**
   * @internal
   */
  [INSPECT_CUSTOM]() {
    return this.diagnosticText
  }
}
