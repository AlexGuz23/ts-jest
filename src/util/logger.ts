import { createLogger, LogContexts, LogLevels } from 'bs-logger'
import { backportTsJestDebugEnvVar } from './backports'

backportTsJestDebugEnvVar()

export const rootLogger = createLogger({
  context: {
    [LogContexts.package]: 'ts-jest',
    [LogContexts.logLevel]: LogLevels.trace,
  },
  targets: process.env.TS_JEST_LOG || undefined,
})
