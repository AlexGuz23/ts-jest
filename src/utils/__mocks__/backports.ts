import * as _backports from '../backports';

export const backportJestConfig: typeof _backports['backportJestConfig'] = val => ({
  ...(val as any),
});
