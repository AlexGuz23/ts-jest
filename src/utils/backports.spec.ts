import { backportJestConfig } from './backports'
import spyThese from '../__helpers__/spy-these'
import set from 'lodash.set'
import { inspect } from 'util'

const consoleSpies = spyThese(console, {
  warn: () => undefined,
})
afterEach(() => {
  consoleSpies.mockReset()
})

describe('backportJestConfig', () => {
  const makeTestsFor = (oldPath: string, newPath: string, values: any[]) => {
    values.forEach(val => {
      let original: any
      beforeEach(() => {
        original = {}
        set(original, oldPath, val)
      })
      describe(`with "${oldPath}" set to ${inspect(val)}`, () => {
        it(`should wran the user`, () => {
          backportJestConfig(original)
          expect(consoleSpies.warn).toHaveBeenCalledTimes(1)
          expect(consoleSpies.warn.mock.calls[0].join(' ')).toMatchSnapshot()
        }) // should warn the user
        it(`should have changed the config correctly`, () => {
          expect(original).toMatchSnapshot('before')
          expect(backportJestConfig(original)).toMatchSnapshot('migrated')
        }) // should have changed the config
      }) // with xxx set to yyy
    }) // for
  } // makeTestsFor

  makeTestsFor('globals.__TS_CONFIG__', 'globals.ts-jest.tsConfig', [
    { foo: 'bar' },
  ])

  makeTestsFor(
    'globals.__TRANSFORM_HTML__',
    'globals.ts-jest.stringifyContentPathRegex',
    [true, false],
  )

  makeTestsFor('globals.ts-jest.tsConfigFile', 'globals.ts-jest.tsConfig', [
    'tsconfig.build.json',
  ])

  makeTestsFor('globals.ts-jest.useBabelrc', 'globals.ts-jest.babelJest', [
    true,
    false,
  ])

  makeTestsFor('globals.ts-jest.babelConfig', 'globals.ts-jest.babelJest', [
    { foo: 'bar' },
  ])

  makeTestsFor('globals.ts-jest.skipBabel', 'globals.ts-jest.babelJest', [
    true,
    false,
  ])
})
