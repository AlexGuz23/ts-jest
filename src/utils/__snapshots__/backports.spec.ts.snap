// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`backportJestConfig with "globals.__TRANSFORM_HTML__" set to false should have changed the config correctly: before 1`] = `
Object {
  "globals": Object {
    "__TRANSFORM_HTML__": false,
  },
}
`;

exports[`backportJestConfig with "globals.__TRANSFORM_HTML__" set to false should have changed the config correctly: migrated 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {},
  },
}
`;

exports[`backportJestConfig with "globals.__TRANSFORM_HTML__" set to true should have changed the config correctly: before 1`] = `
Object {
  "globals": Object {
    "__TRANSFORM_HTML__": true,
  },
}
`;

exports[`backportJestConfig with "globals.__TRANSFORM_HTML__" set to true should have changed the config correctly: migrated 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "stringifyContentPathRegex": "\\\\.html?$",
    },
  },
}
`;

exports[`backportJestConfig with "globals.__TS_CONFIG__" set to { foo: 'bar' } should have changed the config correctly: before 1`] = `
Object {
  "globals": Object {
    "__TS_CONFIG__": Object {
      "foo": "bar",
    },
  },
}
`;

exports[`backportJestConfig with "globals.__TS_CONFIG__" set to { foo: 'bar' } should have changed the config correctly: migrated 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "tsConfig": Object {
        "foo": "bar",
      },
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.babelConfig" set to { foo: 'bar' } should have changed the config correctly: before 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "babelConfig": Object {
        "foo": "bar",
      },
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.babelConfig" set to { foo: 'bar' } should have changed the config correctly: migrated 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "babelJest": Object {
        "foo": "bar",
      },
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.skipBabel" set to false should have changed the config correctly: before 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "skipBabel": false,
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.skipBabel" set to false should have changed the config correctly: migrated 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "babelJest": true,
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.skipBabel" set to true should have changed the config correctly: before 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "skipBabel": true,
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.skipBabel" set to true should have changed the config correctly: migrated 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {},
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.tsConfigFile" set to 'tsconfig.build.json' should have changed the config correctly: before 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "tsConfigFile": "tsconfig.build.json",
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.tsConfigFile" set to 'tsconfig.build.json' should have changed the config correctly: migrated 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "tsConfig": "tsconfig.build.json",
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.useBabelrc" set to false should have changed the config correctly: before 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "useBabelrc": false,
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.useBabelrc" set to false should have changed the config correctly: migrated 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "babelJest": Object {},
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.useBabelrc" set to true should have changed the config correctly: before 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "useBabelrc": true,
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.useBabelrc" set to true should have changed the config correctly: migrated 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "babelJest": true,
    },
  },
}
`;
